### ISSUE TITLE
Issue: Rfe slow execution time with potential failures

### ISSUE THREAD

**Issue by olivierroncalez** [topepo/caret#894]
I've been trying to reproduce the results of feature selection from chapter 19 (Applied Predictive Modeling), but noticed that the R code for Rfe will vary wildly from a few minutes to failure to complete execution (with a manual end to execution after about 30 minutes), even for the same model despite a fixed seed value. 

Usually, the models using the full set of predictors works fine, but some of the models using Rfe need to be stopped as execution never ends. When my machine is restarted and these models are run directly, they will sometimes work and fully execute after 5-8 minutes. It's worth noting that sometimes my R session will never terminate and needs to be manually force closed, while other times it will thorw an error message, "unable to establish connection with R session" after attempting to close execution (and will subsequently restart R). 

Any idea why this may be? The seed is exactly the same and never changes. Code is below. 

As a side note, verbose never prints anything to the console, although I'm not familiar enough with the caret package yet to know if this is expected behavior for Rfe.

## Data Preparation
> library(AppliedPredictiveModeling)
> data(AlzheimerDisease)
> 
> bl <- c("Genotype", "age", "tau", "p_tau", "Ab_42", "male")
> 
> newAssays <- colnames(predictors)
> newAssays <- newAssays[!(newAssays %in% c("Class", bl))]
> 
> predictors$E2 <- predictors$E3 <- predictors$E4 <- 0
> predictors$E2[grepl("2", predictors$Genotype)] <- 1
> predictors$E3[grepl("3", predictors$Genotype)] <- 1
> predictors$E4[grepl("4", predictors$Genotype)] <- 1
> genotype <-  predictors$Genotype
> 
> library(caret)
> set.seed(730)
> split <- createDataPartition(diagnosis, p = .8, list = FALSE)
> 
> adData <- predictors
> adData$Class <- diagnosis
> 
> training <- adData[ split, ]
> testing  <- adData[-split, ]
> 
> predVars <- names(adData)[!(names(adData) %in% c("Class",  "Genotype"))]
> 
> fiveStats <- function(...) c(twoClassSummary(...), defaultSummary(...))
> 
> set.seed(104)
> index <- createMultiFolds(training$Class, times = 5)
> 
> varSeq <- seq(1, length(predVars)-1, by = 2)
> 
> library(doMC)
> registerDoMC(15)
> 
> ctrl <- rfeControl(method = "repeatedcv", repeats = 5,
>                    saveDetails = TRUE,
>                    index = index,
>                    returnResamp = "final",
>                    verbose = TRUE)
> 
> fullCtrl <- trainControl(method = "repeatedcv",
>                          repeats = 5,
>                          summaryFunction = fiveStats,
>                          classProbs = TRUE,
>                          index = index,
>                          verboseIter = TRUE)

## Full model training

> set.seed(721)
> rfFull_time <- system.time(rfFull <- train(training[, predVars],
>                 training$Class,
>                 method = "rf",
>                 metric = "ROC",
>                 tuneGrid = data.frame(mtry = floor(sqrt(length(predVars)))),
>                 ntree = 1000,
>                 trControl = fullCtrl))
> rfFull
> 
> set.seed(721)
> ldaFull_time <- system.time(ldaFull <- train(training[, predVars],
>                  training$Class,
>                  method = "lda",
>                  metric = "ROC",
>                  tol = 1.0e-12,
>                  trControl = fullCtrl))
> ldaFull
> 
> set.seed(721)
> svm_time <- system.time(svmFull <- train(training[, predVars],
>                  training$Class,
>                  method = "svmRadial",
>                  metric = "ROC",
>                  tuneLength = 12,
>                  preProc = c("center", "scale"),
>                  trControl = fullCtrl))
> svmFull
> 
> set.seed(721)
> nbFull_time <- system.time(nbFull <- train(training[, predVars],
>                 training$Class,
>                 method = "nb",
>                 metric = "ROC",
>                 trControl = fullCtrl))
> nbFull
> 
> lrFull <- train(training[, predVars],
>                 training$Class,
>                 method = "glm",
>                 metric = "ROC",
>                 trControl = fullCtrl)
> lrFull
> 
> set.seed(721)
> knnFull_time <- system.time(knnFull <- train(training[, predVars],
>                  training$Class,
>                  method = "knn",
>                  metric = "ROC",
>                  tuneLength = 20,
>                  preProc = c("center", "scale"),
>                  trControl = fullCtrl))
> knnFull

## Rfe Training

> 
> ctrl$functions <- rfFuncs
> ctrl$functions$summary <- fiveStats
> set.seed(721)
> rffRFE_time <- system.time(rfRFE <- rfe(training[, predVars],
>              training$Class,
>              sizes = varSeq,
>              metric = "ROC",
>              ntree = 1000,
>              rfeControl = ctrl))
> rfRFE
> 
> ctrl$functions <- ldaFuncs
> ctrl$functions$summary <- fiveStats
> 
> set.seed(721)
> ldaRFE_time <- system.time(ldaRFE <- rfe(training[, predVars],
>               training$Class,
>               sizes = varSeq,
>               metric = "ROC",
>               tol = 1.0e-12,
>               rfeControl = ctrl))
> ldaRFE
> 
> ctrl$functions <- nbFuncs
> ctrl$functions$summary <- fiveStats
> set.seed(721)
> nbRFE_time <- system.time(nbRFE <- rfe(training[, predVars],
>               training$Class,
>               sizes = varSeq,
>               metric = "ROC",
>               rfeControl = ctrl))
> nbRFE
> 
> ctrl$functions <- caretFuncs
> ctrl$functions$summary <- fiveStats
> 
> cvCtrl <- trainControl(method = "cv",
>                        verboseIter = TRUE,
>                        classProbs = TRUE,
>                        allowParallel = FALSE)
> 
> set.seed(721)
> svmRFE_time <- system.time(svmRFE <- rfe(training[, predVars],
>               training$Class,
>               sizes = varSeq,
>               rfeControl = ctrl,
>               metric = "ROC",
>               ## Now arguments to train() are used.
>               method = "svmRadial",
>               tuneLength = 12,
>               preProc = c("center", "scale"),
>               trControl = cvCtrl))
> svmRFE
> 
> ctrl$functions <- lrFuncs
> ctrl$functions$summary <- fiveStats
> 
> set.seed(721)
> lrRFE_time <- system.time(lrRFE <- rfe(training[, predVars],
>                training$Class,
>                sizes = varSeq,
>                metric = "ROC",
>                rfeControl = ctrl))
> lrRFE
> 
> ctrl$functions <- caretFuncs
> ctrl$functions$summary <- fiveStats
> 
> set.seed(721)
> knnRFE_time <- system.time(knnRFE <- rfe(training[, predVars],
>               training$Class,
>               sizes = varSeq,
>               metric = "ROC",
>               method = "knn",
>               tuneLength = 20,
>               preProc = c("center", "scale"),
>               trControl = cvCtrl,
>               rfeControl = ctrl))
> knnRFE

## Session Info

> sessionInfo()

> R version 3.4.3 (2017-11-30)
> Platform: x86_64-apple-darwin15.6.0 (64-bit)
> Running under: macOS High Sierra 10.13.4
> 
> Matrix products: default
> BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
> LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib
> 
> locale:
> [1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8
> 
> attached base packages:
> [1] parallel  stats     graphics  grDevices utils     datasets  methods   base     
> 
> other attached packages:
> [1] doMC_1.3.5                      iterators_1.0.9                 foreach_1.4.4                   caret_6.0-80                   
> [5] ggplot2_2.2.1                   lattice_0.20-35                 AppliedPredictiveModeling_1.1-7
> 
> loaded via a namespace (and not attached):
>  [1] Rcpp_0.12.17       lubridate_1.7.4    tidyr_0.8.1        class_7.3-14       assertthat_0.2.0   ipred_0.9-6        psych_1.8.4       
>  [8] R6_2.2.2           plyr_1.8.4         magic_1.5-8        stats4_3.4.3       ellipse_0.4.1      pillar_1.2.3       rlang_0.2.1       
> [15] lazyeval_0.2.1     kernlab_0.9-26     rpart_4.1-13       Matrix_1.2-14      splines_3.4.3      CVST_0.2-2         ddalpha_1.3.3     
> [22] gower_0.1.2        stringr_1.3.1      foreign_0.8-70     munsell_0.4.3      broom_0.4.4        compiler_3.4.3     pkgconfig_2.0.1   
> [29] mnormt_1.5-5       dimRed_0.1.0       nnet_7.3-12        tidyselect_0.2.4   tibble_1.4.2       prodlim_2018.04.18 DRR_0.0.3         
> [36] CORElearn_1.52.1   codetools_0.2-15   RcppRoll_0.2.2     withr_2.1.2        dplyr_0.7.5        MASS_7.3-50        recipes_0.1.2     
> [43] ModelMetrics_1.1.0 grid_3.4.3         nlme_3.1-137       gtable_0.2.0       magrittr_1.5       scales_0.5.0       stringi_1.2.2     
> [50] reshape2_1.4.3     bindrcpp_0.2.2     timeDate_3043.102  robustbase_0.93-0  geometry_0.3-6     lava_1.6.1         tools_3.4.3       
> [57] glue_1.2.0         DEoptimR_1.0-8     purrr_0.2.5        sfsmisc_1.1-2      abind_1.4-5        survival_2.42-3    yaml_2.1.19       
> [64] colorspace_1.3-2   cluster_2.0.7-1    bindr_0.1.1

**Comment by olivierroncalez** [comment]
Commenting to update on the issue. I've discovered the connection problems relate to the parallel feature for the rfe and train control. Sometimes R will allow persistent connection and other times it will fail with execution refusing to stop. Is any solution to this known?

**Comment by olivierroncalez** [comment]
Resolved issue: Turns out I had run a Naive Bayes RFE prior and forgot to adjust the control functions to account for SVMs cross fold tuning over all parameters, hence the initial quick time observed. The connection was working the entire time, I was just impatient. 

Few tips as I had seen a prior request for implement a verbose parameter for parallel processing of models in Caret, #85 : making a cluster using `library(doParallel)` and `makeCluster(detectCores() -1, outfile = "")` will allow the verbose arguments of rfeControl and trainControl to print out directly in the consol providing some indication of training time remaining. Interrupting the process when `doParallel` rather than `doMC` will retain the RStudio environment, although the clusters will continue to execute and need to be terminated via the terminal `killall R` on Mac. Interrupting `doMC` restarts the R environment and eliminates all objects. 

I don't know if this was still a desire but I found it useful and was unable to locate this specifically within Caret.
